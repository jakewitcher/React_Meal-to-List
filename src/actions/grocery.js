export const onAddGroceryAsync = (
    {
        name = '',
        items = [],
    } = {}
) => ({
    type: 'ADD_GROCERY_ASYNC',
    grocery: {
        name,
        items,
    }
});

export const onDeleteGroceryAsync = (id = '') => ({
    type: 'DELETE_GROCERY_ASYNC',
    id,
});

export const onEditGroceryAsync = (
    {
        name = '',
        id = '',
        items = [],
    } = {}
) => ({
    type: 'EDIT_GROCERY_ASYNC',
    grocery: {
        id,
        name,
        items,
    }
});

export const onSetGroceryAsync = () => ({
    type: 'SET_GROCERY_ASYNC',
});