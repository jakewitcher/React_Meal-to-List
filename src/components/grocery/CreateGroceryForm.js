import React from 'react';

const CreateGroceryForm = (props) => {
    const {
        selectedMealChange,
        selectedMeal,
        meals,
        groceryName,
        itemsList,
        groceryNameChange,
        addMealToList,
        resetGrocery,
        dispatch,
    } = props.groceryProps;
    const { addGrocery } = props;
    return (
        <div className="form">
            <h2 className="form__header">Create a New Grocery List</h2>
            <div className="form__form">
                <div className="form__nameinputfield">
                    <p>Grocery List Name</p>
                    <input
                        className="form__nameinputfield--input"
                        type="text"
                        value={groceryName}
                        onChange={groceryNameChange}
                    />
                </div>
                <div className="form__iteminputfields">
                    <div>
                        <p>Meals</p>
                        <select
                            className="form__iteminputfields--input form__iteminputfields--dropdown"
                            name="selectedMeal"
                            value={selectedMeal.name}
                            onChange={selectedMealChange}
                        >
                            {
                                meals.mealList.map(meal => <option value={meal.name} key={meal.id}>{meal.name}</option>)
                            }
                        </select>
                    </div>
                    <div className="form__button">
                        <button
                            className="form__button--add"
                            onClick={addMealToList}
                        >
                            Add Meal
                    </button>
                    </div>
                </div>
                <div>
                    <button
                        className="form__button--save"
                        onClick={() => {
                            dispatch(addGrocery({
                                name: groceryName,
                                items: itemsList,
                            }));
                            resetGrocery();
                        }
                        }
                    >
                        Save Grocery List
                </button>
                </div>
            </div>

        </div>
    )
}

export default CreateGroceryForm;