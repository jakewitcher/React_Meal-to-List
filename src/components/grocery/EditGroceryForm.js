import React from 'react';

const EditGroceryForm = (props) => {
    const {
        selectedMealChange,
        selectedMeal,
        meals,
        groceryName,
        itemsList,
        groceryNameChange,
        addMealToList,
        resetGrocery,
        dispatch,
    } = props.groceryProps;
    const { editGrocery, match } = props;
    return (
        <div className="form">
            <h2 className="form__header">Edit Grocery List</h2>
            <div className="form__form">
                <div className="form__nameinputfield">
                    <p>Grocery List Name</p>
                    <input
                        className="form__nameinputfield--input"
                        type="text"
                        value={groceryName}
                        onChange={groceryNameChange}
                    />
                </div>
                <div className="form__iteminputfields">
                    <div>
                        <p>Meals</p>
                        <select
                            className="form__iteminputfields--input form__iteminputfields--dropdown"
                            name="selectedMeal"
                            value={selectedMeal.name}
                            onChange={selectedMealChange}
                        >
                            {
                                meals.mealList.map(meal => <option value={meal.name} key={meal.id}>{meal.name}</option>)
                            }
                        </select>
                    </div>
                    <div className="form__button">
                        <button
                            className="form__button--add"
                            onClick={addMealToList}
                        >
                            Add Meal
                    </button>
                    </div>
                </div>
                <div>
                    <button
                        className="form__button--save"
                        onClick={() => {
                            const id = match.params.groceryId;
                            dispatch(editGrocery({
                                name: groceryName,
                                items: itemsList,
                                id,
                            }));
                            resetGrocery();
                        }
                        }
                    >
                        Save Grocery List
                </button>
                </div>
            </div>

        </div>
    )
}

export default EditGroceryForm;