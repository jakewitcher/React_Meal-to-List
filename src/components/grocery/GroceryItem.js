import React from 'react';

const GroceryItem = (props) =>  {
    const { itemName, amount, unit, id } = props.item;
    return (
        <li className="itemlist__item">
            <div className="item">
                <p>{itemName}</p>
            </div>
            <div className="item">
                <p>{amount}</p>
            </div>
            <div className="item">
                <p>{unit}</p>
            </div>
            {
                !props.isDetails &&
                <div>
                    <i 
                        className="fas fa-trash"
                        onClick={() => props.deleteGroceryItem(id)}    
                    />
                </div>
            }
        </li>
    )
}

export default GroceryItem;