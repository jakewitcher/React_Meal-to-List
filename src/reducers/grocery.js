const groceryDefaultState = {
    groceryList: [],
};

export default (state = groceryDefaultState, action) => {
    switch (action.type) {
        case 'ADD_GROCERY':
            return {
                ...state,
                groceryList: [...state.groceryList, action.grocery]
            };
        default:
            return state;
    }
}