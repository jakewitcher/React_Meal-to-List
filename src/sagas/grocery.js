import { put, takeEvery, select } from "redux-saga/effects";
import database from "../firebase/firebase";

const getUserId = state => state.auth.uid;

const formatItemList = list => {
  return list.reduce((list, item) => {
    list[item.itemName] = {
      amount: item.amount,
      unit: item.unit
    };
    return list;
  }, {});
};

function* addGrocery({ grocery = {} }) {
  const { itemList, name, groceryListMeals } = grocery;
  const items = formatItemList(itemList);
  const meals = groceryListMeals;

  let id;
  const uid = yield select(getUserId);

  yield database
    .ref(`users/${uid}/groceryLists`)
    .push({ name, items, meals })
    .then(ref => {
      id = ref.key;
    });

  yield put({
    type: "ADD_GROCERY",
    grocery: {
      name,
      itemList,
      groceryListMeals,
      id
    }
  });
}

function* editGrocery({ grocery = {} }) {
  const { itemList, name, id, groceryListMeals } = grocery;
  const items = formatItemList(itemList);
  const uid = yield select(getUserId);
  const meals = groceryListMeals;

  yield database
    .ref(`users/${uid}/groceryLists/${id}`)
    .update({ name, items, meals });

  yield put({
    type: "EDIT_GROCERY",
    grocery: {
      name,
      itemList,
      groceryListMeals,
      id
    }
  });
}

function* deleteGrocery({ id = "" }) {
  const uid = yield select(getUserId);

  yield database.ref(`users/${uid}/groceryLists/${id}`).remove();

  yield put({
    type: "DELETE_GROCERY",
    id
  });
}

function* setGrocery() {
  const lists = [];
  const uid = yield select(getUserId);

  yield database
    .ref(`users/${uid}/groceryLists`)
    .once("value")
    .then(snapshot => {
      snapshot.forEach(childSnapshot => {
        const itemList = [];
        childSnapshot.child("items").forEach(item => {
          itemList.push({
            itemName: item.key,
            amount: item.val().amount,
            unit: item.val().unit
          });
        });

        lists.push({
          id: childSnapshot.key,
          name: childSnapshot.val().name,
          itemList,
          groceryListMeals: childSnapshot.val().meals
        });
      });
    });
  yield put({ type: "SET_GROCERY", lists });
}

export function* watchAddGrocery() {
  yield takeEvery("ON_ADD_GROCERY", addGrocery);
}

export function* watchEditGrocery() {
  yield takeEvery("ON_EDIT_GROCERY", editGrocery);
}

export function* watchDeleteGrocery() {
  yield takeEvery("ON_DELETE_GROCERY", deleteGrocery);
}

export function* watchSetGrocery() {
  yield takeEvery("ON_SET_GROCERY", setGrocery);
}
