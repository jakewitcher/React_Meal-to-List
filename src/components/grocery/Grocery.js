import React from 'react';
import { connect } from 'react-redux';
import { deleteGrocery } from '../../actions/grocery';

const Grocery = (props) => {
    const { name, dispatch, editExistingGrocery, id } = props;
    return (
        <div>
            <p>{name}</p>
            <i 
                className="fas fa-edit"
                onClick={() => editExistingGrocery()}
            />
            <i 
                className="fas fa-trash"
                onClick={() => dispatch(deleteGrocery(id))}    
            />
        </div>
    );
}

export default connect()(Grocery);