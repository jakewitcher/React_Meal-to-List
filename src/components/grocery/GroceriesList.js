import React from 'react';
import Grocery from './Grocery';

const GroceriesList = (props) => {
    const { groceryLists } = props;
    return (
        <div>
            {
                groceryLists.map(list => {
                    return <Grocery
                        name={list.name}
                        key={list.id}
                    />
                })
            }
        </div>
    );
}

export default GroceriesList;