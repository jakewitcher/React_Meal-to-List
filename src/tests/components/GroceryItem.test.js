import React from 'react';
import { shallow } from 'enzyme';
import GroceryItem from '../../components/grocery/GroceryItem';
import { groceryItems } from '../fixtures/grocery';

let item, deleteGroceryItem, wrapper;

beforeEach(() => {
    item = {
        itemName: groceryItems[0].itemName,
        amount: groceryItems[0].amount,
        unit: groceryItems[0].unit,
        id: groceryItems[0].id,
    };
    deleteGroceryItem = jest.fn();
    wrapper = shallow(
        <GroceryItem 
            item={item}
            deleteGroceryItem={deleteGroceryItem}
        />
    );
});

test('should render GroceryItem correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should delete grocery item by id', () => {
    wrapper.find('i').simulate('click');
    expect(deleteGroceryItem).toHaveBeenLastCalledWith(item.id);
});