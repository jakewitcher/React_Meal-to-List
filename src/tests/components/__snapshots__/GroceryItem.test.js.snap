// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render GroceryItem correctly 1`] = `
<li
  className="itemlist__item"
>
  <div
    className="item"
  >
    <p>
      chicken
    </p>
  </div>
  <div
    className="item"
  >
    <p>
      2
    </p>
  </div>
  <div
    className="item"
  >
    <p>
      pounds
    </p>
  </div>
  <div>
    <i
      className="fas fa-trash"
      onClick={[Function]}
    />
  </div>
</li>
`;
